sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala

```diff
@@ -42,17 +42,28 @@ case class AnalyzeColumnCommand(
     if (tableMeta.tableType == CatalogTableType.VIEW) {
       throw new AnalysisException("ANALYZE TABLE is not supported on views.")
     }
-    val sizeInBytes = AnalyzeTableCommand.calculateTotalSize(sessionState, tableMeta)
+    val oldRowCount = tableMeta.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L)
+    val oldColStats = tableMeta.stats.map(_.colStats).getOrElse(Map.empty)
+    val newSize = AnalyzeTableCommand.calculateTotalSize(sessionState, tableMeta)
 
     // Compute stats for each column
-    val (rowCount, newColStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+    val (newRowCount, colStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+
+    val newColStats = if (newRowCount == 0) {
+      // No need to keep column stats when there's no data.
+      Map.empty[String, ColumnStat]
+    } else if (newRowCount == oldRowCount) {
```

Comment:
If we can always invalidate the stats, reusing the existing one makes sense when the num of row counts are the same (or the changes are tolerable). 

Could you check how Hive handles this?